### ISSUE TITLE
Flow cannot distinguish method signature in disjoint union

### ISSUE THREAD

**Issue by frontendphil** [facebook/flow#6594]
Working example: https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoALgTwA4FMwDKAlgHYDmMuAKmALxgDeqYYAtgK4zpHaUBcYKAEMYAZ1wAaZmDgkAwgAsh5XAIAUANxHtVYUegBOpMgEo6APjAa4RACaoAvqgw58AWU7ca9Jiw5cefjBDHSkWWUVlMl1NbV0AQQMDIUwAHn0jcnMzWktrO0dnLDwwAAUDOGxRb0JjShoAHzAPAKpnVCh2EgBjblkwAFVeOCFbNQY2T0DJGXklFTAHAXLK6rMmAEgiKDV-bl58WiPggx111A3wuajcNQBtAHIoODgHiTAHgCMhAweAXRM0hYDjAuDE+F8LChEXm0TUTxe31+gKhi0KQA

I wonder why the following code is not type-checking. As far as I've understood disjoint unions the `multiple` property should tell Flow what to expect as `onChange`. However, this doesn't seem to work. I'd like to understand whether this is my fault and how I could get this to type check or whether this is a bug. 

**Code**
```es6
/* @flow */
type SingleT = {
  multiple: false,
  onChange: (value: string) => void
}

type MultiT = {
  multiple: true,
  onChange: (value: Array<string>) => void
}

type PropsT = SingleT | MultiT

function Upload({ multiple, onChange }: PropsT) {
  if(multiple === true) {
    onChange(['foo', 'bar'])
  } else {
    onChange('foobar')
  }
}

```

**Comment by frontendphil** [comment]
Wow, this makes it even weirder :) However, this only works for this example not the actual code where I took this from :/

**Comment by stereobooster** [comment]
Would you mind to post more info or close the issue?

**Comment by frontendphil** [comment]
What kind of info would help?

**Comment by stereobooster** [comment]
@matthewbpt fixed explained why initial code example doesn't work. If this doesn't help solve your issue, please post updated code, otherwise not much we can do.

**Comment by frontendphil** [comment]
Ok, I've updated the example. 

https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoAlgWwA5wE4AuYASgKYCGAxgQDRgDeYAwnDnAHantEC+YUeVmADkeCtWGpUBAJ7ZSYAMrp2AcxikAKmAC8DVGDCYArjALpsGgFz9yMAM6kaBsByYALcmtIB+GwAoANztjUht7AjwVVQBKXQA+MEC4dAATVB4pWXkwAFlTc209ehcTMwtrMEjQ50M3T28-MCCQsLAAQTw8chkAHgiotXi4nUTktIysuQUABUFseyKlaI1tAB88gvRNKUoYcnt7MABVSzhyVLBSAA8CLlSjljZObl65uAXNRJLDMXZU0h4fxxH6GMC9FTYYxEepeVSkHT0YEJKrudD2AB0DVSGg8cNIwL4wHiLkyhhc2NxDXhyNBhnQUH82VIcCgqPRGOw80xsO8YAAhDo9MIoMZ2NR0BxhCCXGCwGICMY8OxZWBMqqGUy0ZiuR9MWVzJYFEK9NVSDK5YYCNrOdyMbyaQBtEVwODCOjCABG5DwwgAujFVXxSA4FHSwdaObqFvb2HjvP4XXBvb7A3KyWrJlIgA

I'm using a `class` component now and also the `onChange` handler is also optional. This is now pretty much exactly the code I'm running. If `onChange` is not optional, it works.

**Comment by frontendphil** [comment]
Absolutely. Can you maybe explain why this has to be done this way? I think it's a bit frustrating that I need to know that I have to write my code in a certain way and do quite weird things in order to make type-checking work. I'll end up with less readable and harder to maintain code for the sake of pleasing flow.

**Comment by stereobooster** [comment]
I will try

```ts
const props = this.props;
props.onChange; // type inferred as ((value: Array<string>) => void) | ((value: string) => void) | void
if (!props.onChange) return;
props.onChange; // type inferred as ((value: Array<string>) => void) | ((value: string) => void)
if(props.multiple === true) {
  // you expect that it will be (value: Array<string>) => void
  // but it is actually remains ((value: Array<string>) => void) | ((value: string) => void)
  // this is basically a bug in Flow algorithm
  props.onChange && props.onChange(['foo', 'bar'])
  // the disappointing part here is that props is inferred as MultiT
}
```

if you will remove `if (!props.onChange) return;` the code will work as you expect.
